daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/UpgradesITDev.scala

```diff
@@ -23,29 +36,265 @@ class UpgradesITDev extends AsyncWordSpec with AbstractScriptTest with Inside wi
+                LedgerClientChannelConfiguration.InsecureDefaults,
+              ),
+            )
+          }
+
+          _ <- traverseSequential(adminClients) { case (ledgerPort, adminClient) =>
+            Future.traverse(deps) { dep =>
+              Thread.sleep(500)
+              println(
+                s"Uploading ${dep.versionedName} to participant on port ${ledgerPort}"
+              )
+              adminClient
+                .uploadDar(dep.path.toFile)
+                .map(_.left.map(msg => throw new Exception(msg)))
+            }
+          }
+
+          // Wait for upload
+          _ <- RetryStrategy.constant(attempts = 20, waitTime = 1.seconds) { (_, _) =>
+            assertDepsVetted(adminClients.head._2, deps)
+          }
+          _ = println("All packages vetted on all participants")
+
+          // Run tests
+          testDar = CompiledDar.read(testDarPath, Runner.compilerConfig(LanguageMajorVersion.V2))
+          _ <- run(
+            clients,
+            QualifiedName.assertFromString(s"${testCase.name}:main"),
+            dar = testDar,
+            enableContractUpgrading = true,
+          )
+        } yield succeed
+      }
+    }
+  }
+
+  private def assertDepsVetted(client: AdminLedgerClient, deps: Seq[DataDep]): Future[Unit] = {
+    client
+      .listVettedPackages()
+      .map(_.foreach { case (participantId, packageIds) =>
+        deps.foreach { dep =>
+          if (!packageIds.contains(dep.packageId))
+            throw new Exception(
+              s"Couldn't find package ${dep.versionedName} on participant $participantId"
+            )
+        }
+      })
+  }
+
+  private val exe = if (sys.props("os.name").toLowerCase.contains("windows")) ".exe" else ""
+  private val damlc = requiredResource(s"compiler/damlc/damlc$exe")
+
+  case class PackageDefinition(
+      name: String,
+      versions: Int,
+      modules: Map[String, String],
+  )
+
+  // TODO[SW] Consider another attempt at using io.circe.generic.auto._
+  // [MA] we make this lazy because we're calling it from the top level before
+  // the entire class has finished loading
+  implicit lazy val decodePackageDefinition: Decoder[PackageDefinition] =
+    new Decoder[PackageDefinition] {
+      final def apply(c: HCursor): Decoder.Result[PackageDefinition] =
+        for {
+          name <- c.downField("name").as[String]
+          versions <- c.downField("versions").as[Int]
+          modules <- c
+            .downField("modules")
+            .as(
+              Decoder
+                .decodeList(Decoder.forProduct2("name", "contents") { (x: String, y: String) =>
+                  (x, y)
+                })
+                .map(Map.from _)
+            )
+        } yield {
+          new PackageDefinition(name, versions, modules)
+        }
+    }
+
+  lazy val packagePattern: Regex = "\\{- PACKAGE *\n((?:.|[\r\n])+?)-\\}".r
+
+  def readPackageDefinitions(fileContent: String): Seq[PackageDefinition] = {
```

Comment:
I think it will more readable if the logic to build those magic dar was extracted in a different object.  
This could be used in other cases like probably `UpgradeSpec` and the integration test. 
If you do not want to do big refactoring maybe you can just put that in another object in the same file with a comment explaining this could maybe reuse somewhere.

Reply:
I've moved the dar creation to a new `DarUtils` file. The extraction is very specific to upgrades so I've kept it in this file but in a separate object